### ISSUE TITLE
[BUG] Memory leak

### ISSUE THREAD

**Issue by SoiledBrush** [seriousm4x/UpSnap#356]
I am not sure what I did wrong while setting up upsnap but after a few days it starts hogging all ram in it's lxc container.

I tried deploying it with docker and running the binary as is.
![Screenshot from 2023-12-09 17-39-50](https://github.com/seriousm4x/UpSnap/assets/43366389/ebadc997-10eb-4d85-a58f-18991c0f3cb0)
on my previous attempt I allocated 2gbs of ram to container.

Here's the current docker-compose:
```
version: "3"
services:
  upsnap:
    container_name: upsnap
    image: ghcr.io/seriousm4x/upsnap:4
    network_mode: host
    restart: unless-stopped
    volumes:
      - ./data:/app/pb_data
    # # To use a non-root user, create the mountpoint first (mkdir data) so that it has the right permission.
    # user: 1000:1000
    environment:
      - TZ=Asia/Almaty # Set container timezone for cron schedules
    #   - UPSNAP_INTERVAL=@every 10s # Sets the interval in which the devices are pinged
      - UPSNAP_SCAN_RANGE=10.0.0.1/24 # Scan range is used for device discovery on local network
    #   - UPSNAP_WEBSITE_TITLE=Custom name # Custom website title
    # # dns is used for name resolution during network scan
    dns:
      - 1.1.1.1
      - 1.0.0.1
    # # you can change the listen ip:port inside the container like this:
    # entrypoint: /bin/sh -c "./upsnap serve --http 0.0.0.0:5000"
    # healthcheck:
    #   test: curl -fs "http://localhost:5000/api/health" || exit 1
    #   interval: 10s
    # # or install custom packages for shutdown
    # entrypoint: /bin/sh -c "apk update && apk add --no-cache <YOUR_PACKAGE> && rm -rf /var/cache/apk/* && ./upsnap serve --http 0.0.0.0:8090"

```

**Comment by seriousm4x** [comment]
That definately shouldn't happen. Did you check with `htop`/`top` if it's actually upsnap thats allocating all that ram?

**Comment by SoiledBrush** [comment]
I would need some memory logging solution for that, as system gets unresponsive when ram get full and I can't check it real time.

**Comment by seriousm4x** [comment]
Can't you just open a shell directly after starting the container and monitor the processes?

**Comment by SoiledBrush** [comment]
Here, after 1 hour uptime:

![Screenshot from 2023-12-14 10-29-34](https://github.com/seriousm4x/UpSnap/assets/43366389/7193e091-35c6-4843-b8eb-8d3337728f07)

15 hours uptime:
![Screenshot from 2023-12-15 00-35-02](https://github.com/seriousm4x/UpSnap/assets/43366389/73e8deac-8a8d-4dbe-8b47-90f09d673d45)

**Comment by seriousm4x** [comment]
Thanks. Are you using any custom shutdown command? Maybe any subprocess does not exit properly.

**Comment by SoiledBrush** [comment]
No, shutdown isn't confugured.
 I was wondering if there is something wrong with my docker-compose.

**Comment by seriousm4x** [comment]
Hm, ok. A missconfigured compose file wouldn't cause a memory leak. It looks good to me.
I will have a look if I can somehow reproduce it. Until then, try updating to 4.2.0 which was just released and see if that magically fixes it.

**Comment by SoiledBrush** [comment]
Ok, if it doesn't. I will also try to to run it bare metal, in case if it's some specific proxmox lxc container bug.

**Comment by SoiledBrush** [comment]
![Screenshot from 2023-12-15 21-06-59](https://github.com/seriousm4x/UpSnap/assets/43366389/0e0ecafb-1e61-43c1-b4e5-b839b47bdab8)
version 4.2
I'll go try it on host itself.

**Comment by SoiledBrush** [comment]
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/ca13922e-2dda-4cbb-a1f9-c38ea8a33004)

Nah, same thing on host machine.
Such a strange behaviour.

**Comment by seriousm4x** [comment]
So, I've setup a proxmox instance with an LXC container and let it run for about 1,5 hours but the ram usage stays the same for me (about 52mb). I've used debian 12, added 1 device with 8 ports that are down and 8 ports that are up. Not sure what else I can try to reproduce it...

![image](https://github.com/seriousm4x/UpSnap/assets/23456686/b922ef6d-aafb-4999-a311-379aedd1c5d0)

**Comment by seriousm4x** [comment]
How many devices do you use in total? And how many ports?
Also what OS are you running on?

**Comment by SoiledBrush** [comment]
It's a proxmox(debian) on a j1800 board. Only ethernet and power plugged in.
I'll go try it on raspberry, brb in 10-15 hours with results.

**Comment by SoiledBrush** [comment]
Ram usage starts ramping after 10-15 hours for me

**Comment by seriousm4x** [comment]
I looked at the code for port checking again. If the port was down, the connection for that request would not be closed. This could lead to the memory leak I guess.

Please try with [4.2.1-beta.1](https://github.com/seriousm4x/UpSnap/releases/tag/4.2.1-beta.1)

**Comment by SoiledBrush** [comment]
Ok, running that one on proxmox
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/056e76d1-5532-4a9d-b273-1cfe0e9b893c)

**Comment by SoiledBrush** [comment]
version 4.2.1-beta.1
![Screenshot from 2023-12-18 08-58-29](https://github.com/seriousm4x/UpSnap/assets/43366389/1a4f0fb6-8dc3-4125-9f86-01fd82b91736)

**Comment by SoiledBrush** [comment]
Here's logs of some errors that started appearing during the run:
```
Dec 18 02:02:00 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:00 cronjobs.go:101: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:01 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:00 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:03 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:02 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:04 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:03 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:05 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:04 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:06 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:02:05 wake.go:16: Wake triggered for proxmox main
Dec 18 02:02:09 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:08 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:10 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:09 cronjobs.go:101: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:02:26 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:02:25 wake.go:16: Wake triggered for proxmox main
Dec 18 02:02:34 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:02:32 wake.go:16: Wake triggered for mosk nas
Dec 18 02:02:41 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:02:39 wake.go:16: Wake triggered for proxmox main
Dec 18 02:02:55 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:02:53 wake.go:16: Wake triggered for proxmox main
Dec 18 02:02:56 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:52 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:02:58 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:54 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:00 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:55 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:01 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:58 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:03 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:02:58 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:06 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:01 cronjobs.go:119: proxmox main%!(EXTRA string=not online after 2 min)
Dec 18 02:03:07 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:06 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:07 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:07 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:08 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:07 cronjobs.go:119: proxmox main%!(EXTRA string=not online after 2 min)
Dec 18 02:03:08 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:08 cronjobs.go:119: proxmox main%!(EXTRA string=not online after 2 min)
Dec 18 02:03:12 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:03:11 wake.go:16: Wake triggered for mosk nas
Dec 18 02:03:12 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:03:12 cronjobs.go:119: mosk nas%!(EXTRA string=not online after 2 min)
Dec 18 02:03:15 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:03:14 wake.go:16: Wake triggered for proxmox main
Dec 18 02:03:27 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:03:25 wake.go:16: Wake triggered for proxmox main
Dec 18 02:03:46 upsnap-114 upsnap[1084]: [INFO] 2023/12/18 02:03:44 wake.go:16: Wake triggered for mosk nas
Dec 18 02:05:12 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:05:09 cronjobs.go:35: context deadline exceeded; failed query: SELECT `devices`.* FROM `devices`
Dec 18 02:05:13 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:05:11 cronjobs.go:78: context deadline exceeded; failed query: SELECT {{_collections}}.* FROM `_collections` WHERE [[id]] = {:id} OR LOWER([[name]])={:name} LIMIT 1
Dec 18 02:05:20 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:05:17 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:05:24 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:05:21 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
Dec 18 02:05:26 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:05:18 cronjobs.go:35: context canceled; failed query: SELECT `@@__missing_devices`.* FROM `@@__missing_devices`
```
and right before stopping service:
```
Dec 18 02:59:02 upsnap-114 upsnap[1084]: [ERROR] 2023/12/18 02:58:59 cronjobs.go:127: Failed to save record: sql: database is closed
```

**Comment by seriousm4x** [comment]
Looks like the database is very slow or cannot be accessed properly. That's why the queries are failing.
Where is your database stored? On an internal SSD or HDD? On a network share?

**Comment by SoiledBrush** [comment]
I think the last one is just and error from stopping the systemd service I created for upsnap. Everything is stored on ssd.


I think I've found the problem. 
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/41405068-b6de-4cee-ab2f-99c0459e6d68)
I have disabled the wake cron option in settings. It's been 9 hours and I don't see any increase in ram usage. 
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/588ba4df-53dd-4e32-8c51-69898ed3bd36)

**Comment by seriousm4x** [comment]
Yup, I was able to reproduce it. The easiest way was to set the wake cron to `@every 1s` on one or more devices, and memory usage would skyrocket. This was also the case for every other cronjob, such as shutdowns and even pings (only noticeable if the interval setting changed multiple times).

The issue was basically that a new Go routine was created for every time any device record in the database changed. It's fixed now.

Thanks for finding it!

Now, cronjobs don't create new Go routines and are only restarted if one of the following database fields changes: `wake_cron`, `wake_cron_enabled`, `shutdown_cron`, `shutdown_cron_enabled`.
Additionally, wake/shutdown crons are not executed anymore when the device is in a `pending` state.

Could you verify if it's fixed for you as well by trying [4.2.1-beta.2](https://github.com/seriousm4x/UpSnap/releases/tag/4.2.1-beta.2)? (Docker build still pending)

**Comment by SoiledBrush** [comment]
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/ea99e0d9-7ab1-4a13-8c1f-3c4977961404)
Yeah, I've set 2 devices to wake every second on beta 2. And, unlike previous version, there is no increase in ram usage.
![image](https://github.com/seriousm4x/UpSnap/assets/43366389/c86a09fa-0e87-4aa5-b7b2-833682f4c9ed)

Good job!
